### ISSUE TITLE
Class level validations on configuration don't report the path of the failing property element

### ISSUE THREAD

**Issue by BenjaminTanguay** [smallrye/smallrye-config#935]
I have a quarkus project that uses smallrye-config and hibernate-validator to handle the project configuration and its validation. Because the project is configuration driven, I want to support the developpers using it as much as possible by providing useful error messages when a configuration is invalid. 

I have an interface that represents a portion of the configuration with some validation logic in it:
```
// @CrossValidation(crossValidationType = CrossValidationType.THEN_RESPOND_MESSAGE_FIELD_COHERENT_WITH_RESPONSE_STRATEGY,
//        message = "Message field must be defined when response strategy is set to MESSAGE")
public interface ThenRespond {
    @NotNull
    @WithConverter(ResponseStrategy.ResponseStrategyConverter.class)
    ResponseStrategy responseStrategy();
    Optional<@Min(100) @Max(599) String> code();

    Optional<String> message();

    @AssertTrue(message = "Message field must be defined when response strategy is set to MESSAGE")
    private boolean isResponseStrategyAndMessageFieldCoherent() {
        return responseStrategy() != ResponseStrategy.MESSAGE ||
                message().isPresent();
    }
}
```

Here, `@Min` and `@Max` refer to custom function level annotations that validate the bounds of the property found under code(). Those annotations yield the expected result. When my configuration contains a code property that is out of bound, I get the following error message:
io.smallrye.config.ConfigValidationException: Configuration validation failed:
	**_universal-stub.stubs[2].defaultRule.thenRespond.code_** @Max validation failed. Field must have a lower or equal value to 599. Value found: 700

Notice here I get the full path of the failing configuration element.

This interface also contains a class level cross-validation between two fields. As a test, I've implemented this cross validation in two different ways: through a custom @CrossValidation annotation at the top of the class and through a private method wired with the @AssertTrue annotation. 

Both mecanisms produce a similar error message:

io.smallrye.config.ConfigValidationException: Configuration validation failed:
	responseStrategyAndMessageFieldCoherent Message field must be defined when response strategy is set to MESSAGE

Notice here how there is no path pointing towards the failing property in the configuration. I would instead have expected an error message having a shape similar to this:

io.smallrye.config.ConfigValidationException: Configuration validation failed:
	universal-stub.stubs[2].defaultRule.thenRespond.responseStrategyAndMessageFieldCoherent Message field must be defined when response strategy is set to MESSAGE

I debugged the production of the ConfigValidationException to see where this behavior found its source and to see whether I could change it somehow. The source of the problem is found in the interface io.smallrye.config.validator.BeanValidationConfigValidator. Since I use quarkus as a framework and it handles the integration of the configuration related libraries for me, I don't have any mecanism to override anything in this interface. This is a problem that belongs to the framework and not this library. However, I find that the default implementation of the methods in this interface could easily be improved to provide the complete path towards the failing element in the configuration. 

The following methods are at the source of the problem described above.

```
    @Override
    default void validateMapping(
            final Class<?> mappingClass,
            final String prefix,
            final Object mappingObject)
            throws ConfigValidationException {

        final List<Problem> problems = new ArrayList<>();
        final ConfigMappingInterface mappingInterface = ConfigMappingInterface.getConfigurationInterface(mappingClass);
        if (mappingInterface != null) {
            validateMappingInterface(mappingInterface, prefix, mappingInterface.getNamingStrategy(), mappingObject, problems);
        } else {
            validateMappingClass(mappingObject, problems);
        }

        if (!problems.isEmpty()) {
            throw new ConfigValidationException(problems.toArray(ConfigValidationException.Problem.NO_PROBLEMS));
        }
    }
```

```
    default void validateMappingInterface(
            final ConfigMappingInterface mappingInterface,
            final String currentPath,
            final NamingStrategy namingStrategy,
            final Object mappingObject,
            final List<Problem> problems) {

        for (Property property : mappingInterface.getProperties()) {
            validateProperty(property, currentPath, namingStrategy, mappingObject, false, problems);
        }

        validateMappingClass(mappingObject, problems);
    }
```

```
    default void validateMappingClass(final Object mappingObject, final List<Problem> problems) {
        final Set<ConstraintViolation<Object>> violations = getValidator().validate(mappingObject);
        for (ConstraintViolation<Object> violation : violations) {
            problems.add(violation.getPropertyPath().toString().isEmpty() ? new Problem(violation.getMessage())
                    : new Problem(violation.getPropertyPath() + " " + violation.getMessage()));
        }
    }
```

The method `validateMappingClass` is the one responsible of handling class level validations. None of its parameters contains any information of the path of the class we're currently validating. The call to `violation.getPropertyPath()` yields nothing useful in that regard. 

The method `validateMappingClass` is called by the methods `validateMapping` and `validateMappingInterface`. Of the two, the method `validateMappingInterface` contains the parameter `currentPath` that contains the desired information. The method `validateMapping` can calls both `validateMappingInterface` and `validateMappingClass`. It contains the parameter `prefix` that is used to seed the `currentPath` when `validateMappingInterface` is called. 

In other words, in all the contexts where `validateMappingClass` is called, the calling entity has the path information available. 

I suggest the following breaking change: the signature of the method `validateMappingClass` should change to:
`void validateMappingClass(final Object mappingObject, final List<Problem> problems, final String currentPath)` and the path information should be used to produce more useful errors for class constraints violations.

**Comment by radcortez** [comment]
Thanks for finding the issue. Would you like to submit a PR to fix it? Thanks!

**Comment by BenjaminTanguay** [comment]
Sure! I'll try to write something this week.

Just to reiterate, since I'll modify a public interface, this is a breaking change. You'll have to consider that the next time you make a release for the library.

**Comment by BenjaminTanguay** [comment]
#938
